politeiad/backendv2/tstorebe/plugins/pi/hooks.go

```diff
@@ -390,11 +395,246 @@ func (p *piPlugin) voteSummary(token []byte) (*ticketvote.SummaryReply, error) {
+// token.
+func (p *piPlugin) commentVoteAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
+	// Decode payload
+	var v comments.Vote
+	err := json.Unmarshal([]byte(payload), &v)
+	if err != nil {
+		return err
+	}
+	if !p.isAncestorOf(latestAuthorUpdate.CommentID, v.CommentID,
+		cs) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeVoteStatusInvalid),
+			ErrorContext: "vote has ended; comment votes are allowed only on " +
+				"the latest author update thread",
+		}
+	}
+	return nil
+}
+
+// commentNewAllowedOnApprovedProposal verifies that the given new comment
+// is allowed on the approved proposal associated with the provided
+// token.
+func (p *piPlugin) commentNewAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
+	// Decode payload
+	var n comments.New
+	err := json.Unmarshal([]byte(payload), &n)
+	if err != nil {
+		return err
+	}
+	// Option 1: the new comment is a new author update
+	//
+	// Get proposal authour to ensure comment's author is
+	// the proposal's author.
+	recordAuthorID, err := p.recordAuthor(token)
+	if err != nil {
+		return err
+	}
+	if n.UserID == recordAuthorID &&
+		n.ExtraData != "" && n.ExtraDataHint != "" &&
+		n.ParentID == 0 {
+		switch n.ExtraDataHint {
+		case pi.ProposalUpdateHint:
+			// Decode comment extra data
+			var pum pi.ProposalUpdateMetadata
+			err = json.Unmarshal([]byte(n.ExtraData), &pum)
+			if err != nil {
+				return err
+			}
+			// Verify update title
+			if !p.updateTitleIsValid(pum.Title) {
+				return backend.PluginError{
+					PluginID:     pi.PluginID,
+					ErrorCode:    uint32(pi.ErrorCodeUpdateTitleInvalid),
+					ErrorContext: p.titleRegexp.String(),
+				}
+			}
+			// New valid author update
+			return nil
+		default:
+			return backend.PluginError{
+				PluginID:  pi.PluginID,
+				ErrorCode: uint32(pi.ErrorCodeExtraDataHintInvalid),
+			}
+		}
+	}
+	// Option 2: the new comment is a reply on one of the replies
+	// on the latest author update or on the update itself.
+	//
+	// New comment is a new thread but not a valid author update
+	if n.ParentID == 0 {
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeVoteStatusInvalid),
+			ErrorContext: "vote has ended; comments are locked",
+		}
+	}
+	// New comment is a reply, ensure it's on the latest
+	// author update thread.
+	if !p.isAncestorOf(latestAuthorUpdate.CommentID, n.ParentID,
+		cs) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeVoteStatusInvalid),
+			ErrorContext: "vote has ended; replies are allowed only on " +
+				"the latest author update thread",
+		}
+	}
+	return nil
+}
+
+// writesAllowedOnApprovedPropsoal verifies that the given comment write
+// is allowed on the approved proposal associated with the provided token.
```

Comment:
"approved" is ambigious in this context.  It could mean the proposal was made public or the vote was approved.  I would update this comment to make it more clear.

Apply to the other functions that reference an `ApprovedProposal` too.